### PR TITLE
[rllib] Port remainder of algorithms to build_trainer() pattern

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Ports DQN, DDPG, QMIX, APEX variants, IMPALA, APPO, MARWIL to use the build_trainer() pattern for their trainers.

We still have to port the policies for some of these algorithms, but that can be done in a separate PR.

## Related issue number

https://github.com/ray-project/ray/issues/4788

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by ericl** [comment]
jenkins retest this please

File: python/ray/rllib/agents/trainer_template.py [hunk 1]
```diff
@@ -81,18 +126,38 @@ def _train(self):
...
+            state = Trainer.__getstate__(self)
+            state.update(self.state)
+            return state
+
+        def __setstate__(self, state):
+            Trainer.__setstate__(self, state)
+            self.state = state
+
     @staticmethod
     def with_updates(**overrides):
```
Comment (richardliaw) [inline python/ray/rllib/agents/trainer_template.py]: what is this method? can you add a docstring?

File: python/ray/rllib/agents/dqn/dqn.py [hunk 2]
```diff
@@ -128,202 +122,169 @@
...
+        schedule_max_timesteps=config["schedule_max_timesteps"],
+        beta_annealing_fraction=config["beta_annealing_fraction"],
+        final_prioritized_replay_beta=config["final_prioritized_replay_beta"],
+        prioritized_replay_eps=config["prioritized_replay_eps"],
+        train_batch_size=config["train_batch_size"],
+        sample_batch_size=config["sample_batch_size"],
+        **config["optimizer"])
+
+
+def check_config_and_setup_param_noise(config):
```
Comment (richardliaw) [inline python/ray/rllib/agents/dqn/dqn.py]: Can you add a docstring for this?

**Review by richardliaw — APPROVED** [review]
Mostly looks good; some parts are not easy to know what's going on.

**Pull request merged**